File: app/src/main/java/com/boardgamegeek/filterer/GeekRankingFilterer.java
package com.boardgamegeek.filterer;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import com.boardgamegeek.R;
import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.util.MathUtils;
import com.boardgamegeek.util.StringUtils;

import java.util.Locale;

public class GeekRankingFilterer extends CollectionFilterer {
	public static final int MIN_RANGE = 1;
	public static final int MAX_RANGE = 2000;

	private int min;
	private int max;
	private boolean includeUnranked;

	public GeekRankingFilterer(Context context) {
		super(context);
	}

	public GeekRankingFilterer(Context context, int min, int max, boolean includeUnranked) {
		super(context);
		this.min = min;
		this.max = max;
		this.includeUnranked = includeUnranked;
	}

	@Override
	public void setData(@NonNull String data) {
		String[] d = data.split(DELIMITER);
		min = d.length > 0 ? MathUtils.constrain(StringUtils.parseInt(d[0], MIN_RANGE), MIN_RANGE, MAX_RANGE) : MIN_RANGE;
		max = d.length > 1 ? MathUtils.constrain(StringUtils.parseInt(d[1], MAX_RANGE), MIN_RANGE, MAX_RANGE) : MAX_RANGE;
		includeUnranked = d.length > 2 && (d[2].equals("1"));
	}

	@Override
	public int getTypeResourceId() {
		return R.string.collection_filter_type_geek_ranking;
	}

	@NonNull
	@Override
	public String flatten() {
		return String.valueOf(min) + DELIMITER + String.valueOf(max) + DELIMITER + (includeUnranked ? "1" : "0");
	}

	public int getMax() {
		return max;
	}

	public int getMin() {
		return min;
	}

	public boolean includeUnranked() {
		return includeUnranked;
	}

	@Override
	public String getDisplayText() {
		return describeRange(R.string.unranked_abbr);
	}

	@Override
	public String getDescription() {
		return context.getString(R.string.ranked) + " " + describeRange(R.string.unranked);
	}

	private String describeRange(@StringRes int unrankedResId) {
		String text;
		if (min >= MAX_RANGE) {
			text = String.format(Locale.getDefault(), "%,d", MAX_RANGE) + "+";
		} else if (min == max) {
			text = String.format(Locale.getDefault(), "%,d", max);
		} else {
			text = String.format(Locale.getDefault(), "%,d-%,d", min, max);
		}
		if (includeUnranked) text += String.format(" (+%s)", context.getString(unrankedResId));
		return "#" + text;
	}

	@Override
	public String getSelection() {
		String selection = min >= MAX_RANGE ?
			String.format("%s>=?", Games.GAME_RANK) :
			String.format("(%1$s>=? AND %1$s<=?)", Games.GAME_RANK);
		if (includeUnranked) selection += String.format(" OR %1$s=0 OR %1$s IS NULL", Games.GAME_RANK);
		return selection;
	}

	@Override
	public String[] getSelectionArgs() {
		return min >= MAX_RANGE ?
			new String[] { String.valueOf(MAX_RANGE) } :
			new String[] { String.valueOf(min), String.valueOf(max) };
	}
}

File: app/src/main/java/com/boardgamegeek/filterer/GeekRankingFilterer.kt
package com.boardgamegeek.filterer

import android.content.Context
import androidx.annotation.StringRes
import com.boardgamegeek.R
import com.boardgamegeek.extensions.IntervalDelegate
import com.boardgamegeek.model.CollectionItem
import com.boardgamegeek.model.Game
import java.util.Locale

class GeekRankingFilterer(context: Context) : CollectionFilterer(context) {
    var min by IntervalDelegate(LOWER_BOUND, LOWER_BOUND, UPPER_BOUND)
    var max by IntervalDelegate(UPPER_BOUND, LOWER_BOUND, UPPER_BOUND)
    var includeUnranked = false

    override val typeResourceId = R.string.collection_filter_type_geek_ranking

    override fun inflate(data: String) {
        val d = data.split(DELIMITER)
        min = d.getOrNull(0)?.toIntOrNull() ?: LOWER_BOUND
        max = d.getOrNull(1)?.toIntOrNull() ?: UPPER_BOUND
        includeUnranked = d.getOrNull(2) == "1"
    }

    override fun deflate() = "$min$DELIMITER$max$DELIMITER${if (includeUnranked) "1" else "0"}"

    override val iconResourceId: Int
        get() = R.drawable.ic_baseline_emoji_events_24

    override fun chipText() = describe(R.string.unranked_abbr)

    override fun description() = "${context.getString(R.string.ranked)} ${describe(R.string.unranked)}"

    private fun describe(@StringRes unrankedResId: Int): String {
        val unranked = if (includeUnranked) " (+${context.getString(unrankedResId)})" else ""
        return "#${describeRange()}$unranked"
    }

    fun describeRange() = when {
        max == UPPER_BOUND -> String.format(Locale.getDefault(), "%,d+", min)
        min == max -> String.format(Locale.getDefault(), "%,d", max)
        min == LOWER_BOUND -> String.format(Locale.getDefault(), "<%,d", max)
        else -> String.format(Locale.getDefault(), "%,d-%,d", min, max)
    }

    override fun filter(item: CollectionItem): Boolean {
        return when {
            item.rank == Game.RANK_UNKNOWN -> includeUnranked
            max == UPPER_BOUND -> item.rank >= min
            min == LOWER_BOUND -> item.rank <= max
            min == max -> item.rank == min
            else -> item.rank in min..max
        }
    }

    companion object {
        const val LOWER_BOUND = 1
        const val UPPER_BOUND = 2000
    }
}
